MainWindow.xaml.cs

```diff
@@ -270,5 +268,54 @@ private void ButtonFill_Click(object sender, RoutedEventArgs e)
         {
             PanZoom.ViewModel.FillRegion();
         }
+
+        #region Closing
+
+        public bool UnsavedChanges
+        {
+            get => unsavedChanges;
+            set
+            {
+                unsavedChanges = value;
+
+                if (unsavedChanges)
+                {
+                    InputPack.Textures.CollectionChanged -= ChangesObserved;
+                    InputPack.HostDevices.CollectionChanged -= ChangesObserved;
+                    InputPack.EmulatedDevices.CollectionChanged -= ChangesObserved;
+                }
+                else
+                {
+                    InputPack.Textures.CollectionChanged += ChangesObserved;
+                    InputPack.HostDevices.CollectionChanged += ChangesObserved;
+                    InputPack.EmulatedDevices.CollectionChanged += ChangesObserved;
+                }
+            }
+        }
+        private bool unsavedChanges = false;
+
+        private void ChangesObserved(object sender, System.Collections.Specialized.NotifyCollectionChangedEventArgs e) => UnsavedChanges = true;
+
+        private void Window_Closing(object sender, System.ComponentModel.CancelEventArgs e)
+        {
+            if (unsavedChanges)
+            {
+                MessageBoxResult MessageResult;
+                MessageResult = MessageBox.Show("unsaved changes are lost, do you want to save?", "unsaved changes!", MessageBoxButton.YesNoCancel);
+                switch (MessageResult)
+                {
+                    case MessageBoxResult.Cancel:
+                        e.Cancel = true;
+                        break;
+                    case MessageBoxResult.Yes:
+                        SaveAsData_Click(sender, new RoutedEventArgs());
```

Comment:
I think this should be `SaveData_Click` if the document has been saved already, otherwise `SaveAsData_Click`

Reply:
Looks like just changing this to `SaveData_Click` is good enough

Reply:
that is right, perhaps most logical for the user.